### PR TITLE
Add Line provider

### PR DESCRIPTION
**Pull request by linmasaki**
- Add Line provider and its unit test.
- Update README.md and solution file.
- Add the documentation.

This provider is one of the popular social chat platforms in Taiwan & Japan.

### CODE REVIEW

File: src/AspNet.Security.OAuth.Line/AspNet.Security.OAuth.Line.csproj [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>netcoreapp3.1</TargetFrameworks>
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Line/AspNet.Security.OAuth.Line.csproj]: Our `dev` branch now targets `net5.0`.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
+        {
+        }
+
+        protected override async Task HandleChallengeAsync(AuthenticationProperties properties)
+        {
+            await base.HandleChallengeAsync(properties);
+        }
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs]: This isn't changing any behaviour so can be removed.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+                ["redirect_uri"] = redirectUri,
+                ["state"] = Options.StateDataFormat.Protect(properties),
+                ["scope"] = FormatScope(),
+                ["prompt"] = Options.Prompt ? "consent" : "none"
+            });
+
+        protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync()
+        {
+            return await base.HandleRemoteAuthenticateAsync();
+        }
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs]: Same here, it's not needed.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            });
+
+            using var response = await Backchannel.SendAsync(request, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving an access token: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                    /* Status: */ response.StatusCode,
+                    /* Headers: */ response.Headers.ToString(),
+                    /* Body: */ await response.Content.ReadAsStringAsync());
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs]: Nit: alignment (like the similar code in `CreateTicketAsync()`).

```suggestion
                                /* Status: */ response.StatusCode,
                                /* Headers: */ response.Headers.ToString(),
                                /* Body: */ await response.Content.ReadAsStringAsync());
```

File: src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+                throw new HttpRequestException("An error occurred while retrieving user information.");
+            }
+
+            using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
+            var principal = new ClaimsPrincipal(identity);
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement);
+            context.RunClaimActions();
+
+            // When the email address is not public, retrieve it from the emails endpoint if the user:email scope is specified.
+            if (!string.IsNullOrEmpty(Options.UserEmailsEndpoint) && Options.Scope.Contains("email"))
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs:104]: The `"email"` claim could go into the relevant constants class, then users can reference it to add the claim to the options, and the code here can use it to reference it.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            });
+
+            using var response = await Backchannel.SendAsync(request, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogWarning("An error occurred while retrieving the email address associated with the logged in user: " +
+                                  "the remote server returned a {Status} response with the following payload: {Headers} {Body}.",
+                    /* Status: */ response.StatusCode,
+                    /* Headers: */ response.Headers.ToString(),
+                    /* Body: */ await response.Content.ReadAsStringAsync());
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs]: Nit: alignment.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace AspNet.Security.OAuth.Line
+{
+    public class LineAuthenticationHandler : OAuthHandler<LineAuthenticationOptions>
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs:24]: This class needs a number of changes applied for .NET 5.0.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationOptions.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+            ClaimsIssuer = LineAuthenticationDefaults.Issuer;
+            CallbackPath = LineAuthenticationDefaults.CallbackPath;
+
+            AuthorizationEndpoint = LineAuthenticationDefaults.AuthorizationEndpoint;
+            TokenEndpoint = LineAuthenticationDefaults.TokenEndpoint;
+            UserInformationEndpoint = LineAuthenticationDefaults.UserInformationEndpoint;
+
+            Scope.Add("profile");
+            Scope.Add("openid");
+            Scope.Add("email");
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Line/LineAuthenticationOptions.cs]: Could use the constant as mentioned in the handler comments.

Comment (martincostello) [reply src/AspNet.Security.OAuth.Line/LineAuthenticationOptions.cs]: Does this need to be the default, and instead something that's opted into?

File: src/AspNet.Security.OAuth.Line/AspNet.Security.OAuth.Line.csproj (hunk 1 above)
Comment (linmasaki) [reply src/AspNet.Security.OAuth.Line/AspNet.Security.OAuth.Line.csproj]: Ok, I have done it.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs (hunk 2 above)
Comment (linmasaki) [reply src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs]: Oh! I forgot to remove it.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs (hunk 3 above)
Comment (linmasaki) [reply src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs]: Done.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs (hunk 4 above)
Comment (linmasaki) [reply src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs]: Sorry, I had ignored to align the code and modified it in a new commit.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs (hunk 6 above)
Comment (linmasaki) [reply src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs]: Done.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs (hunk 5 above)
Comment (linmasaki) [reply src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs:104]: I remove it from the default scopes then users can add it to the scope optionally if need.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs (hunk 7 above)
Comment (linmasaki) [reply src/AspNet.Security.OAuth.Line/LineAuthenticationHandler.cs:24]: Done it.

File: src/AspNet.Security.OAuth.Line/LineAuthenticationOptions.cs (hunk 8 above)
Comment (linmasaki) [reply src/AspNet.Security.OAuth.Line/LineAuthenticationOptions.cs]: I update it and see above. Thank you for your suggestions.

Comment (linmasaki) [reply src/AspNet.Security.OAuth.Line/LineAuthenticationOptions.cs]: By the way, how can I make a version to target netcoreapp3.1(use the branch rel/3.1.7)?

**Comment by linmasaki** [comment]
By the way, how can I make a version to target netcoreapp3.1(use the branch rel/3.1.7)?

**Comment by martincostello** [comment]
We weren't particularly planning on supporting both 3.1 and 5.0 versions of all the providers. The last release was a bit of an anomaly as the 5.0 release wasn't yet merged to the dev branch so we had both still active while people were still adding new providers.

_If_ we were to do that going forwards, then we'd need to make an alternate dev branch for 3.1, and then you'd have to make a second PR to adding the same provider targeting 3.1 to that branch.

Making all contributors do two PRs to add a new provider to keep things in sync doesn't really seem fair, and isn't something we have the bandwidth to do for them for every PR being raised to this repo.

**Comment by linmasaki** [comment]
OK, thank you for your response. If I want to see it on Nuget that including both 3.1 and 5.0, how can I do it(since most of my production sites still target 3.1)?

**Comment by martincostello** [comment]
Without us committing to supporting both versions, the only way would be for you to separately and independently publish a package with a different name/ID that targeted 3.1.

Otherwise you’d have to support both yourself from an independent repository and not commit this code to our repository.

**Comment by linmasaki** [comment]
Um... understand, so at this time, this repository only accepts versions above 5.0, right? I have to create another package to support 3.1 if I want to use it on net core app 3.1.

**Comment by martincostello** [comment]
That's the current position, yes. We don't really have the bandwidth to maintain two parallel branches for 3.1 and 5.0 support.

**Comment by linmasaki** [comment]
Ok, I got it. Thank you for your prompt comment.

**Comment by linmasaki** [comment]
All suggestions have been updated. Thanks for advising those.

**Comment by martincostello** [comment]
Are you happy to just have the NuGet package targeting `net5.0` and being published from this repo then?

**Comment by linmasaki** [comment]
Sure, I just confused why not target 3.1 first, but you have explained it. Thanks :smile:

**Comment by martincostello** [comment]
Great thanks - I'll look at merging the current open PRs and publishing a 5.0.2 release with them to NuGet.org some time this week.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Thanks again for your contribution - the Line provider is now available from NuGet.org: https://www.nuget.org/packages/AspNet.Security.OAuth.Line/

**Pull request merged**